src/gwt/src/org/rstudio/core/client/command/AppCommand.java

```diff
@@ -100,19 +114,79 @@ public void onVisibleChanged(AppCommand command)
          parentToolbar_.invalidateSeparators();
       }
 
-      private final AppCommand command_;
+      protected final AppCommand command_;
       private boolean synced_ = true;
-      private HandlerRegistration handlerReg_;
-      private HandlerRegistration handlerReg2_;
-      private Toolbar parentToolbar_;
+      protected HandlerRegistration handlerReg_;
+      protected HandlerRegistration handlerReg2_;
+      protected Toolbar parentToolbar_;
+   }
+
+   private class CommandSourceColumnToolbarButton extends CommandToolbarButton
+   {
+      public CommandSourceColumnToolbarButton(String buttonLabel,
+                                              String buttonTitle,
+                                              ImageResourceProvider imageResourceProvider,
+                                              ClickHandler clickHandler,
+                                              AppCommand command,
+                                              SourceColumn column)     {
+         super(buttonLabel, buttonTitle, imageResourceProvider, clickHandler, command, false);
+         column_ = column;
+      }
+
+      @Override
+      protected void onAttach()
+      {
+         super.onAttach();
+
+         handlerReg_ = command_.addEnabledChangedHandler(this);
```

Comment:
Instead of keeping track of each registration, use a `HandlerRegistrations` to remember them and then unload them all at once. https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/core/client/HandlerRegistrations.java